qiskit_experiments/framework/composite/batch_experiment.py

```diff
@@ -127,3 +127,69 @@ def _remap_qubits(self, circuit, qubit_mapping):
         new_circuit.metadata = circuit.metadata
         new_circuit.append(circuit, qubit_mapping, list(range(num_clbits)))
         return new_circuit
+
+    def _run_jobs(self, circuits: List[QuantumCircuit], **run_options) -> List[Job]:
+        if len(circuits) == 0:
+            return []
+
+        if self.experiment_options.separate_jobs:
+            try:
+                is_first_time = False
+                # The local metadata is a truncation of the original composite metadata,
+                # such that the current experiment (self) is at the head of the metadata.
+                # In the root experiment (at the beginning of the reucrsion) none of the circuits
+                # contains the local_metadata attribute; after the first iteration they all do.
+                if not hasattr(circuits[0], "local_metadata"):
+                    is_first_time = True
+                    for circ in circuits:
+                        circ.local_metadata = circ.metadata
+
+                # A dictionary that maps sub-experiments to their circuits
+                circs_by_subexps = {}
+                for circ in circuits:
+                    # For batch experiments the composite index is always a list of length 1,
+                    # because unlike parallel experiment, each circuit originates from a single
+                    # sub-experiment.
+                    circ_index = circ.local_metadata["composite_index"][0]
+                    if circ_index in circs_by_subexps:
```

Comment:
You can also use [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict) if you want.